File: three_sum/brute_force.py
import unittest
import timeit

class ThreeSum:

    def __init__(self, array):
      self.array = array
      self.n = len(self.array)

    def count(self):
      count = 0
      for i in range(self.n):
        for j in range(i+1, self.n):
          for k in range(j+1, self.n):
            if (self.array[i] + self.array[j] + self.array[k] == 0):
              count += 1 
      return count

class ThreeSumTest(unittest.TestCase):
    
  def test_count(self):
    arr1 = [-3, -2, -1, 0, 1, 2, 3]
    obj = ThreeSum(arr1)
    self.assertEqual(obj.count(), 5)

    arr2 = [3, 2, 1, 0, 1, 2, 3]
    obj = ThreeSum(arr2)
    self.assertEqual(obj.count(), 0)

if __name__ == '__main__':
    setup = '''
from __main__ import ThreeSum
import random

array = random.sample(range(-1000, 1000), 100)
'''
    print(max(timeit.repeat("ThreeSum(array).count()", setup= setup, number= 10)))
    unittest.main()

File: three_sum/brute_force.rb
require 'test/unit'
require 'benchmark'

class ThreeSum
 
  def initialize array=[]
    @array = array
    @n = (@array.size) -1
  end

  def count
    count = 0
    (0..@n).to_a.each do |i|
      j = i+1
      (j..@n).to_a.each do |j|
        k = j+1
        (k..@n).to_a.each do |k|
          count += 1 if (@array[i] + @array[j] + @array[k] == 0)
        end
      end
    end
    return count
  end

end

# Check performance
BM_ARRAY = (-10000..10000).map { rand }

Benchmark.bm do |x|
  x.report { 50.times do;  ThreeSum.new(BM_ARRAY.sample(100)); end }
end

# Test object
class ThreeSumTest < Test::Unit::TestCase
  
  def test_count
    arr1 = [-3, -2, -1, 0, 1, 2, 3]
    obj = ThreeSum.new(arr1)
    assert_equal(5, obj.count())

    arr2 = [3, 2, 1, 0, 1, 2, 3]
    obj = ThreeSum.new(arr2)
    assert_equal(0, obj.count())
  end

end
